plugins/autowarmth.koplugin/suntime.lua

```diff
@@ -242,33 +306,117 @@ function SunTime:getTimeDiff(height)
     return 12/math.pi * acos(val)
 end
 
--- get time for a certain height
--- set hour to 6 for rise or 18 for set
--- result rise or set time
+-- Get time for a certain height
+-- Set hour near to expected time
+-- Sed after_noon to true, if sunset is wanted
+-- Result rise or set time
 --        nil sun does not reach the height
-function SunTime:calculateTime(height, hour)
-    hour = hour or 12
+function SunTime:calculateTime(height, hour, after_noon)
     local dst = self.date.isdst and 1 or 0
     local timeDiff = self:getTimeDiff(height, hour)
     if not timeDiff then
         return
     end
 
     local local_correction = self.time_zone - self.pos.longitude*12/math.pi + dst - self.zgl
-    if hour < 12 then
+    if not after_noon then
         return 12 - timeDiff + local_correction
     else
         return 12 + timeDiff + local_correction
     end
 end
 
+-- If height is nil, use newly calculated self.eod
 function SunTime:calculateTimeIter(height, hour)
-    return self:calculateTime(height, hour)
+    local after_noon = hour > 12
+
+    self:initVars(hour) -- calculate self.eod
+    hour = self:calculateTime(height or self.eod, hour, after_noon)
+    if hour ~= nil then
+        self:initVars(hour)  -- calculate self.eod
+        hour = self:calculateTime(height or self.eod, hour, after_noon)
+    end
+    return hour
+end
+
+function SunTime:calculateNoon()
+    self:initVars(12)
+    if self.pos.latitude >= 0 then -- northern hemisphere
+        if math.pi/2 - self.pos.latitude + self.decl > self.eod then
+            local dst = self.date.isdst and 1 or 0
+            local local_correction = self.time_zone - self.pos.longitude*12/math.pi + dst - self.zgl
+            return 12 + local_correction
+        end
+    else -- sourthern hemisphere
+        if math.pi/2 + self.pos.latitude - self.decl > self.eod then
+            local dst = self.date.isdst and 1 or 0
+            local local_correction = self.time_zone - self.pos.longitude*12/math.pi + dst - self.zgl
+            return 12 + local_correction
+        end
+    end
+end
+
+function SunTime:calculateMidnight()
+    -- 24 is the midnight at the end of the current day,
+    -- 00 would be the beginning of the day
+    self:initVars(24)
+    if self.pos.latitude >= 0 then -- northern hemisphere
+        if math.pi/2 - self.pos.latitude - self.decl > self.eod then
+            local dst = self.date.isdst and 1 or 0
+            local local_correction = self.time_zone - self.pos.longitude*12/math.pi + dst - self.zgl
+            return 24 + local_correction
+        end
+    else -- southern hemisphere
+        if math.pi/2 + self.pos.latitude + self.decl > self.eod then
+            local dst = self.date.isdst and 1 or 0
+            local local_correction = self.time_zone - self.pos.longitude*12/math.pi + dst - self.zgl
+            return 24 + local_correction
+        end
+    end
+
 end
 
+--[[--
+Calculates the ephemeris ant twilight times
```

Comment:
Typo probably, to include in future improvements. ;-)

Btw, one day is a bit short, please try to wait at least two unless it's small or urgent.

Reply:
Sorry, next time I will wait longer.